Reduce HGlobal allocation in CallResult scenario

Akarinnnnn opened the pull request:
Reduced HGlobal allocation times in CallResult scenario by cache IntPtr buffer .

Changed .NET Standard project to make use of new(.NET 5+) API, Unity will also get benefit from it. Compatibility to Framework and old Unity is handled by conditional compilcation.

Akarinnnnn committed:
Reduced HGlobal allocation times in CallResult scenario.

Changed .NET Standard project to make use of new(.NET 5+) API, Unity will also get benefit from it. Compatibility to Framework and old Unity is handled by conditional compilcation.

Akarinnnnn committed:
Fixed an HGlobal leak in CallbackDispatcher.

Leaked HGlobal is allocated by member initialization statement, removed it.

rlabrecque commented on com.rlabrecque.steamworks.net/Runtime/CallbackDispatcher.cs:
```diff
@@ -48,6 +49,9 @@ public static void ExceptionHandler(Exception e) {
 		private static IntPtr m_pCallbackMsg;
 		private static int m_initCount;
+		// 4096 * 3 is enough for most api calls, 
+		private static int s_currentCallResultBufferSize = 4096 * 3;
```
The comment on this scales me a little bit

rlabrecque commented on com.rlabrecque.steamworks.net/Runtime/CallbackDispatcher.cs:
```diff
@@ -69,12 +73,54 @@ internal static void Shutdown() {
+		private static IntPtr CheckCallResultBufferSize(uint requiredBufferSize)
+        {
+            if (s_currentCallResultBufferSize >= requiredBufferSize)
+                return s_pCallResultBuffer; // buffer is enough, this happens mostly
```
Can we wrap all of these single line if statements in { } ?

Also comment is semi-cut off

Akarinnnnn replied:
Sure, style issues should be fixed at final stage. Now I want to fix code design first.

Akarinnnnn replied:
Now all re-formatted. I think add a .editorconfig will help to enforce style issues.

rlabrecque commented:
Any objections @JamesMcGhee @yaakov-h ?

yaakov-h commented:
I'm a little bit concerned about thread safety and concurrency re. the static buffer, and also that there is no limit on the buffer until hitting INT_MAX and then crashing, particularly when this buffer stays around for the lifetime of the application.

If use of ArrayPool/ObjectPool is available or similar then it might be better to reuse that system.

Also throwing raw IntPtrs just kind of gives me the heebie-jeebies in general.

Akarinnnnn commented:
> I'm a little bit concerned about thread safety and concurrency re. the static buffer, and also that there is no limit on the buffer until hitting INT_MAX and then crashing, particularly when this buffer stays around for the lifetime of the application.

Your concerns is right, but I see some usage of static shared buffers on CallbackMsg_t, so this might be acceptable for now. Buffer shrinking is also a good point, I plan to fix it by a counter about buffer too large. And also, the exception thrown by reaching INT_MAX is handled and will only skip particular result, no need to worry about crashing whole application.

> If use of ArrayPool/ObjectPool is available or similar then it might be better to reuse that system.

Considering we still support some clients that don't have native `ArrayPool` and `Span` support, if we want to do pooling, we will have to write one ourself, which will add some burden to maintenance I afraid.

JamesMcGhee commented:
Overall no objections to the concept not where I can run and test myself but read over the change set.

As to the Thread Safty aspect
Implementing developer's responsibility, a Unity dev should naturally always register on main that said a developer that has worked with Unreal will know that they have callbacks coming in on Steam's own thread so it is on you in each delegate to handle directing to the desired thread i.e. it is an implementing developer's problem.

On the overflow topic, it should have belts and braces to check for and handle overflow, there are a few ways you could implement recovering space it should be a non-issue there aren't that many CallResult calls that should be called so frequently that it would be a problem. That said there should never be an Access Violation or Overlow that is unhandled either not in C# anyway

Akarinnnnn committed:
Encapsule shared buffer into a class to handle threading stuff

Akarinnnnn commented:
Wait, let me do some test first.

Akarinnnnn committed:
Fixed call-result buffer allocation logic, added a debug purpose configuration.

Akarinnnnn commented:
Functional test under .NET 8 and .NET Standard 2.1 passed.

Akarinnnnn commented:
Functional test code below, call me for making project setup public.
```csharp
// See https://aka.ms/new-console-template for more information
using Steamworks;

Console.WriteLine("Hello, World!");
CancellationTokenSource loopCts = new();

ulong modFAGE_id = 1579583001;

try
{

	SteamAPI.Init();

	Thread dispatcher = new(() =>
	{
		while (!loopCts.IsCancellationRequested)
			SteamAPI.RunCallbacks();
	});

	dispatcher.IsBackground = true;
	dispatcher.Start();

	// Add test code here
	
	CallResult<SteamUGCQueryCompleted_t>[] completionHolder = [.. GenerateReceivers(10, null!)];

	for (int i = 0; i < completionHolder.Length; i++)
	{
		var queryHandle = SteamUGC.CreateQueryUGCDetailsRequest([new(modFAGE_id + (uint)i)], 1);
		var queryApiCallHandle = SteamUGC.SendQueryUGCRequest(queryHandle);
		completionHolder[i].Set(queryApiCallHandle); 
	}



	SpinWait spinner = new();
	while (Sync.CompletedJobsCount != 10)
	{
		spinner.SpinOnce();
	}
}
finally
{
	loopCts.Cancel();
	SteamAPI.Shutdown();
}

static IEnumerable<CallResult<SteamUGCQueryCompleted_t>> GenerateReceivers(int amount, Barrier? completionBarrier)
{
	// completionBarrier.AddParticipants(amount);

	for (int i = 0; i < amount; i++)
	{
		yield return new((r, f) =>
		{
			if (f)
			{
				Console.WriteLine("API Invocation IO Failure");

				Interlocked.Increment(ref Sync.CompletedJobsCount);
				return;
			}

			var completedHandle = r.m_handle;
			if (!SteamUGC.GetQueryUGCResult(completedHandle, 0, out var details))
			{
				Console.WriteLine("GetResult() generic failure");

                Interlocked.Increment(ref Sync.CompletedJobsCount);
				return;
			}

			if (!SteamUGC.GetQueryUGCStatistic(completedHandle, 0, EItemStatistic.k_EItemStatistic_NumComments, out ulong commentsCount))
			{
                Console.WriteLine("GetStatistic() generic failure");

                Interlocked.Increment(ref Sync.CompletedJobsCount);
				return;
            }

            Console.WriteLine($"Title: {details.m_rgchTitle}, URL: {details.m_rgchURL}, Comments count: {commentsCount}");

            Interlocked.Increment(ref Sync.CompletedJobsCount); 
			return;
        });
	}
}


static class Sync
{
    public static volatile int CompletedJobsCount = 0;
}
```

Akarinnnnn committed:
Make code style in CallbackDispatcher.cs consist.

Akarinnnnn committed:
Keep same indentation style as old CallbackDispatcher.cs.

Akarinnnnn committed:
Merge branch 'master' into improve-dispatcher

Akarinnnnn committed:
Remove "Any CPU" and "Deb/Rel" in solution config

Akarinnnnn committed:
Abandon .NET 8 to fix nuget pack workflow for standalone SDK project.

Introduce `TargetFrameworks` have critial effect to build process, which results breaking batch-build step of nuget pack.

Considering side effect above introduced by adding .NET 8 support, to maintain packageing functions, abandon .NET 8 now.

Akarinnnnn committed:
Make call-result buffer size more resaonable.

In addition, original buggy size rounding is replaced with a more plain implementation.

Akarinnnnn commented:
I feel this version is more satisfied, ready to review.
